ui/js/models/repository.js

```diff
@@ -141,13 +153,31 @@ treeherder.factory('ThRepositoryModel', [
                         });
 
                         _.each(data, addRepoAsUnwatched);
+
+                        // This needs to be done before `setCurrent` because
+                        // `setCurrent` overwrites the entire listing
+                        // with only the default repo
+                        if (options.watchRepos) {
+                            var storedWatched = loadWatchedRepos();
+                        }
+
                         if (options.name) {
                             setCurrent(options.name);
                         }
                         if (options.watchRepos) {
-                            var storedWatched = loadWatchedRepos();
-                            if (_.isArray(storedWatched) &&
-                                _.contains(storedWatched, options.name)) {
+                            if (_.isArray(storedWatched)) {
+
+                                // To get the current repo to display first, we must
+                                // ensure it's added to the array last, as per normal user interaction
+                                if (_.contains(storedWatched, options.name)) {
+                                    storedWatched = _.without(storedWatched, _.findWhere(storedWatched, options.name));
+                                }
+                                unwatchRepo(options.name);
+
+                                // Add the repos in reverse order, like the user would (oldest to newest)
+                                storedWatched.reverse();
```

Comment:
Since we store the desired order in localStorage, adding the items to the watch list in reverse order "simulates" how a user would add them (oldest first, newest last), and thus newest added to the front each time.